### PR TITLE
docs: add http api rfc

### PR DESCRIPTION
**Pull request by dl239**
## Links

- [Full Rendered Proposal](https://github.com/dl239/rfcs/blob/main/fedb/api-server.md)

---

**Important: Do NOT comment on this PR. Please use the discussion thread linked above to provide feedback, as it provides branched discussions that are easier to follow. This also makes the edit history of the PR clearer.**

### CODE REVIEW

File: fedb/api-server.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
+- Start Date: 2021-04-23
+- Target Major Version: (2.2.0)
+- Reference Issues: 
+- Implementation PR: 
+
+# Add API Server
+
+# Summary
+
+HTTP interface is one of the most friendly interface and many developer like using http interface to develop their program. HTTP API also help us to show demo simply. So we want to support uing http api to access FEDB.
```
Comment (lotabout) [inline fedb/api-server.md]: HTTP API -> REST API?

File: fedb/api-server.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+## Interface design
+HTTP URL: http://ip:port/sql
+
+HTTP Method: POST
+
+### Put
+request
+```
+{
+    "method": "put",
```
Comment (lotabout) [inline fedb/api-server.md]: Current description is a bit hard to determine the interface, which is expected?

1. `PUT /sql/table/{table_name}/db/{db_name}`
2. `PUT /sql?table={table_name}&db={db_name}`
3. or put all the params into request body?

File: fedb/api-server.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+# Detailed design
+
+
+## Interface design
+HTTP URL: http://ip:port/sql
+
+HTTP Method: POST
+
+### Put
```
Comment (lotabout) [inline fedb/api-server.md:35]: Any batch put API?

File: fedb/api-server.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+}
+```
+response
+```
+{
+    "code": 0,
+    "msg": "ok",
+    "data": [{"field1" : "value1"}, {"field1": "value2"}]
```
Comment (lotabout) [inline fedb/api-server.md]: What's the expected size of output? Extracting schema might be an alternative.

File: fedb/api-server.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+                                google::protobuf::Closure* done) {
+            // parse request attachment
+            ...
+        }
+    };
+    ```
+3. Dispatch and forward request
+    
+    ```
+    if (method == "put") {
```
Comment (lotabout) [inline fedb/api-server.md]: So this is not REST?... And there is only one entry point for all methods? Feels like RPC...

File: fedb/api-server.md (hunk 3 above)
Comment (dl239) [reply fedb/api-server.md:35]: sdk/server does not support batch put now

File: fedb/api-server.md (hunk 4 above)
Comment (dl239) [reply fedb/api-server.md]: it seems a good idea when there are many records in output.

File: fedb/api-server.md (hunk 5 above)
Comment (dl239) [reply fedb/api-server.md]: It can also config a mapping form method to service interface.

File: fedb/api-server.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+HTTP interface is one of the most friendly interface and many developers like using http interface to develop their program. REST API also help us to show demo simply. So we want to support uing http api to access FEDB.
+
+# Detailed design
+
+
+## Interface design
+
+### Put
+reqeust url: http://ip:port/sql/put/{db_name}/{table_name}
```
Comment (aceforeverd) [inline fedb/api-server.md]: 一般 API 描述：
+ URL
+ METHOD
+ HEADERS (if has)
+ REQUEST BODY (if has)
+ STATUS CODE
+ RESPONSE BODY

其他通用的
+ error code

**Comment by wuyou10206** [comment]
delete and put may have security problems, which need attention

File: fedb/api-server.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+{
+    "code": 0,
+    "msg": "ok",
+    "data":{
+        "name": "procedure_name",
+        "procedure": "xxxxx",
+        "common_col": ["field1", "field2"],
+        "input_schema": [{"name":"field1", "type":"bool"}, {"name":"filed1", "type":"int32"}],
+        "output_schema": [{"name":"field1", "type":"bool"}, {"name":"field2", "type":"int32"}],
+        "tables": ["table1", "table2"]
```
Comment (imotai) [inline fedb/api-server.md:111]: add a `sql` field to show the content of sql script

Comment (imotai) [reply fedb/api-server.md:111]: the `output_schema` also has `common_col`

File: fedb/api-server.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+}
+```
+response:
+```
+{
+    "code": 0,
+    "msg": "ok",
+    "data": {
+        "schema": [{"field1":"bool"}, {"field2":"int32"}],
+        "data": [["value1", "value2"], [...]]
```
Comment (imotai) [inline fedb/api-server.md]: The `response` should include common_cols and  none_common_cols

File: fedb/api-server.md (hunk 7 above)
Comment (dl239) [reply fedb/api-server.md:111]: > add a `sql` field to show the content of sql script

`procedure` field show the content of sql script

Comment (dl239) [reply fedb/api-server.md:111]: > the `output_schema` also has `common_col`

ok

File: fedb/api-server.md (hunk 8 above)
Comment (dl239) [reply fedb/api-server.md]: ok

File: fedb/api-server.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+      public:
+        bool Init() {
+            // init the sdk
+            ...
+        }
+        virtual void ProcessSQL(google::protobuf::RpcController* controller,
+                                const HttpRequest* request,
+                                HttpResponse* response,
+                                google::protobuf::Closure* done) {
+            // parse url and dispach
```
Comment (imotai) [inline fedb/api-server.md]: dispatch

File: fedb/api-server.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+1. Design proto file.  
+    ```
+    option cc_generic_services = true;
+
+    message HttpRequest { };
+    message HttpResponse { };
+
+    service APIService {
+        rpc ProcessSQL(HttpRequest) returns (HttpResponse);
```
Comment (imotai) [inline fedb/api-server.md:134]: `ProcessSQL` sounds strange, Can we just call it `Process`?

**Review by imotai — APPROVED** [review]
LGTM

**Review by vagetablechicken — APPROVED** [review]
LGTM

**Review by aceforeverd — APPROVED** [review]

**Review by tobegit3hub — APPROVED** [review]
LGTM

File: fedb/api-server.md [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+### Execute Procedure 
+reqeust url: http://ip:port/dbs/{db_name}/procedures/{procedure_name}  
+http method: POST  
+request body: 
+```
+{
+    "common_cols":["value1", "value2"],
+    "input": [["value0", "value3"],["value0", "value3"]],
+    "need_schema": true
```
Comment (vagetablechicken) [inline fedb/api-server.md]: note that we don't have to add `need_schema`, default is false？
So is `common_cols`.

File: fedb/api-server.md [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+### HTTP Status 
+API Server will return 200 in most case and store the error in the code of reponse's json.
+
+### Put
+reqeust url: http://ip:port/dbs/{db_name}/tables/{table_name}  
+http method: PUT  
+request body: 
+```
+{
+    "value": [[
```
Comment (vagetablechicken) [inline fedb/api-server.md:41]: note we don't support multi put, even the req format supports.

File: fedb/api-server.md [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+            // parse url and dispatch
+        }
+    };
+    ```
+3. Add mapping
+    
+    ```
+    if (server.AddService(&api_svc,
+                      brpc::SERVER_DOESNT_OWN_SERVICE,
+                      "/dbs/*  => Process" != 0) {
```
Comment (vagetablechicken) [inline fedb/api-server.md]: The impl is "/* => Process", which one needs fix?

Comment (dl239) [reply fedb/api-server.md]: done

**Pull request merged**